tee/prototype/epochproc.go

```diff
@@ -0,0 +1,397 @@
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	err "perun.network/go-perun/pkg/errors"
+
+	"github.com/perun-network/erdstall/tee"
+)
+
+// TODO: order function after call-stack please.
+
+func (e *Enclave) epochProcessor(
+	verifiedBlocks <-chan *tee.Block,
+	txs <-chan *tee.Transaction,
+) error {
+	var (
+		depositEpoch = NewEpoch(0)
+		txEpoch      *Epoch
+		exitEpoch    *Epoch
+		phaseShift   = make(chan struct{})
+		errg         = err.NewGatherer()
+	)
+	// push first epoch
+	e.epochs.Push(depositEpoch)
+
+	for {
+		select {
+		case <-e.quit:
+			return errg.Err()
+		default:
+		}
+		done := make(chan exitersSet)
+
+		errg.Go(func() error {
+			return e.depositExitRoutine(verifiedBlocks, done, depositEpoch, exitEpoch)
+		})
+
+		errg.Go(func() error {
+			return e.txRoutine(phaseShift, done, txs, txEpoch)
+		})
+
+		select {
+		case <-phaseShift:
+			e.shiftEpochs(&depositEpoch, &txEpoch, &exitEpoch)
+		case <-errg.Failed():
+			return errg.Err()
+		}
+	}
+}
+
+// TODO: racing transactions.
+func (e *Enclave) txRoutine(
+	phaseShift chan<- struct{},
+	exits <-chan exitersSet,
+	txs <-chan *tee.Transaction,
+	txEpoch *Epoch,
+) error {
+	var stagedTxs txCache
+	for {
+		bpCache, err := e.generateBalanceProofs(txEpoch)
+		if err != nil {
+			return fmt.Errorf("generating balance proofs: %w", err)
+		}
+		select {
+		case exiters := <-exits:
+			txEpoch, bpCache, ptxs := prune(txEpoch, stagedTxs, bpCache, exiters)
+			bpCache, err = e.applyEpochTx(txEpoch, ptxs, bpCache)
+			if err != nil {
+				return fmt.Errorf("adjusting Epoch %v Balances: %w", txEpoch.Number, err)
+			}
+			e.balanceProofs <- getBalProofs(bpCache)
+			phaseShift <- struct{}{} // phase done, stop processing TXs and signal epoch shift.
+			return nil
+		case tx := <-txs:
+			stagedTxs = cacheTx(stagedTxs, tx)
+		}
+	}
+}
+
+// getBalProofs retrieves a slice of `*tee.BalanceProof`s from the balance proof cache.
+func getBalProofs(bps balProofCache) []*tee.BalanceProof {
+	proofs := make([]*tee.BalanceProof, len(bps))
+	i := 0
+	for _, bp := range bps {
+		proofs[i] = bp
+	}
+	return proofs
+}
+
+// prune prunes a given slice of transactions for a given set of addresses, s.t.
+// no element of the given set is either a recipient or sender in a transaction,
+// it also makes sure to remove exited parties from the txEpoch and the
+// balance proof cache.
+func prune(epoch *Epoch,
+	txs txCache,
+	bpc balProofCache,
+	forS []common.Address,
+) (*Epoch, balProofCache, []*tee.Transaction) {
+	for _, m := range forS {
+		delete(epoch.balances, m) // remove member from txEpoch
+		delete(bpc, m)            // remove member from balProofCache
+
+		sTxs, sOk := txs.senders[m]
+		rTxs, rOk := txs.recipients[m]
+		if !sOk && !rOk {
+			continue
+		}
+		for _, tx := range sTxs {
+			delete(txs.txs, tx.Hash())
+		}
+		for _, tx := range rTxs {
+			delete(txs.txs, tx.Hash())
+		}
+	}
+	prunedTxs := make([]*tee.Transaction, 0, len(txs.txs))
```

Comment:
```suggestion
	prunedTxs := make([]*tee.Transaction, len(txs.txs))
```
or the loop will panic

Reply:
of course! Got too late :see_no_evil: